crates/nu-utils/standard_library/std.nu

```diff
@@ -18,88 +18,123 @@ export-env {
+#    ·           ─┬─
+#    ·            ╰── They are not equal: 3 != 1
+#    ╰────
+#
+#
+# 👇👇👇 BE CAREFUL! 👇👇👇
+# ❯ assert ( 1 == 1.0) # passes
+# ❯ assert eq 1 1.0
 # Error:
-#   × left is not equal to right
-#    ╭─[entry #12:5:1]
-#  5 │     if not $cond {
-#  6 │         error make {msg: $msg}
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
+#   × Assertion failed.
+#    ╭─[entry #16:1:1]
+#  1 │ assert eq 1 1.0
+#    ·           ──┬──
+#    ·             ╰── Different types cannot be equal: int <-> float.
 #    ╰────
+# 
 # ```
-export def "assert eq" [left: any, right: any] {
-    _assert (($left | describe) == ($right | describe)) $"left and right operand of `assert eq` should have the same type"
-    _assert ($left == $right) "left is not equal to right"
+export def "assert eq" [left: any, right: any, message?: string] {
+    let left_type = ($left | describe)
+    let right_type = ($right | describe)
+    let left_start = (metadata $left).span.start
+    let right_end = (metadata $right).span.end
+
+    if ($left_type != $right_type) {
+        _assertion-error $left_start $right_end $"Different types cannot be equal: ($left_type) <-> ($right_type)." $message
+    }
+    if ($left != $right) {
+        _assertion-error $left_start $right_end $"They are not equal: ($left) != ($right)" $message
+    }
 }
 
 # ```nushell
-# >_ assert_ne $a "a string"
+# ❯ assert ne 1 3
+# ❯ assert ne 42 42
 # Error:
-#   × left and right operand of `assert eq` should have the same type
-#    ╭─[entry #12:5:1]
-#  5 │     if not $cond {
-#  6 │         error make {msg: $msg}
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
+#   × Assertion failed.
+#    ╭─[entry #23:1:1]
+#  1 │ assert ne 42 42
+#    ·           ──┬──
+#    ·             ╰── They both are 42
 #    ╰────
+# 
 #
-# >_ assert_ne $a 1
-# >_ assert_ne $a 3
+# 👇👇👇 BE CAREFUL! 👇👇👇
+# ❯ assert ( 1 != "a string" ) # passes
+# ❯ assert ne 1 "a string"
 # Error:
-#   × left is equal to right
-#    ╭─[entry #12:5:1]
-#  5 │     if not $cond {
-#  6 │         error make {msg: $msg}
-#    ·         ─────┬────
-#    ·              ╰── originates from here
-#  7 │     }
+#   × Assertion failed.
+#    ╭─[entry #20:1:1]
+#  1 │ assert ne 1 "a string"
+#    ·           ──────┬─────
+#    ·                 ╰── They are not equal, although they have different types: int <-> string.
 #    ╰────
 # ```
-export def "assert ne" [left: any, right: any] {
-    _assert (($left | describe) == ($right | describe)) $"left and right operand of `assert eq` should have the same type"
-    _assert ($left != $right) "left is equal to right"
+export def "assert ne" [left: any, right: any, message?: string] {
+    let left_type = ($left | describe)
+    let right_type = ($right | describe)
+    let left_start = (metadata $left).span.start
+    let right_end = (metadata $right).span.end
+
+    if (($left | describe) == ($right | describe)) {
```

Comment:
Maybe check for inequality here so you don't have to have a no-op block after if?
```suggestion
    if (($left | describe) != ($right | describe)) {
```